### PR TITLE
DATAUP-389 - specify job requirements in run_job[_batch]

### PR DESCRIPTION
**Pull request by MrCreosote**
# Description of PR purpose/changes

What it says on the tin.

Next steps in a future PR:
- in the Impl.py file, pass in an `as_admin` parameter so it's actually possible to specify job requirements at the API level
- add integration tests

# Jira Ticket / Github Issue #
- [x] Added the Jira Ticket to the title of the PR e.g. (DATAUP-69 Adds a PR template)

# Testing Instructions
* Details for how to test the PR: 
- [x] Tests pass in Github Actions and locally 
- [x] Changes available by spinning up a local test suite

# Dev Checklist:

- [x] My code follows the guidelines at https://sites.google.com/truss.works/kbasetruss/development
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [n/a] I have made corresponding changes to the documentation
- [?] My changes generate no new warnings - 3k warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [n/a] Any dependent changes have been merged and published in downstream modules
- [x] I have run Black and Flake8 on changed Python Code manually or with git precommit (and the Github Actions build passes)

# Updating Version and Release Notes (if applicable)

- [n/a] [Version has been bumped](https://semver.org/) for each release
- [n/a] [Release notes](/RELEASE_NOTES.md) have been updated for each release (and during the merge of feature branches)

### CODE REVIEW

File: lib/execution_engine2/sdk/EE2Runjob.py [hunk 1]
```diff
@@ -331,47 +336,125 @@ def run_batch(
...
+                    wsid=job.get(_WORKSPACE_ID),
+                    source_ws_objects=job.get(_SOURCE_WS_OBJECTS),
+                )
+            except IncorrectParamsException as e:
+                self._rethrow_incorrect_params_with_error_prefix(e, pre)
             if has_parent_job and job.get(_PARENT_JOB_ID):
-                pre = f"Job #{i + 1}: b" if len(jobs) > 1 else "B"
                 raise IncorrectParamsException(
-                    f"{pre}atch jobs may not specify a parent job ID"
+                    f"{pre}batch jobs may not specify a parent job ID"
```
Comment (MrCreosote) [inline lib/execution_engine2/sdk/EE2Runjob.py:457]: Sorry AJ, maintaining proper capitalization was just not worth the effort

Comment (ialarmedalien) [reply lib/execution_engine2/sdk/EE2Runjob.py:457]: 😱

File: test/tests_for_sdkmr/EE2Runjob_test.py [hunk 2]
```diff
@@ -482,17 +577,85 @@ class and its respective composed classes, and we don't reproduce all the error
...
+        IncorrectParamsException("job_requirements must be a mapping"),
+    )
+    _run_and_run_batch_fail_illegal_arguments(
+        {
+            "method": "foo.bar",
+            "job_requirements": {
+                "bill_to_user": {
+                    "Bill": "$3.78",
+                    "Boris": "$2.95",
+                    "AJ": "$1.32",
```
Comment (ialarmedalien) [inline test/tests_for_sdkmr/EE2Runjob_test.py:591]: I'm a cheap date, obvs

File: test/tests_for_sdkmr/EE2Runjob_test.py [hunk 3]
```diff
@@ -230,27 +293,59 @@ def test_run_as_admin():
...
+        "client_group": client_group,
+        "client_group_regex": True,
+        "debug_mode": True,
+    }
+    jrr.get_requirements_type.return_value = RequirementsType.BILLING
+    req_args = {
+        "cpus": cpus,
+        "memory_MB": mem,
+        "disk_GB": disk,
+        "client_group": client_group,
```
Comment (ialarmedalien) [inline test/tests_for_sdkmr/EE2Runjob_test.py:309]: scrolling through this file, there are a lot of places where dictionaries with these four args are created -- if I were writing this code, I'd either set up a default with every field populated and then create a new dict by merging that with whatever field you're changing, or have a dict with a subset of the fields that don't change, and merge it with extra params for the other fields. In either situation, you'd be creating a copy of the original dictionary before merging with the other params.

File: test/tests_for_sdkmr/EE2Runjob_test.py (hunk 2 above)
Comment (MrCreosote) [reply test/tests_for_sdkmr/EE2Runjob_test.py]: Fixed

File: test/tests_for_sdkmr/EE2Runjob_test.py (hunk 3 above)
Comment (MrCreosote) [reply test/tests_for_sdkmr/EE2Runjob_test.py]: Fixed

**Review by ialarmedalien — APPROVED** [review]
🥇

**Pull request merged**